[Core] Fix resource validation for existing cluster and refactor

Michaelvll opened the pull request:
Fixes #1258 

This PR makes validation of the resources based on the offering for each region/zone, only when the resources is `launchable` and is being launched by the optimizer. This is because if the cluster already exists, the user can specify some combination of the accelerator / accelerator count that does not exist in the specified region/zone.

For example: aws only has A100:8 in a region, but the user can `sky exec cluster_name --region region_name --cloud aws --gpus A100:1` to schedule the job on the existing cluster. Our current master PR will block this behavior, as the A100:1 is not offered by the region.

The validation of the Resources for existing cluster will be done by https://github.com/skypilot-org/skypilot/blob/8d3fce2a166a1d83b741cdabdba152b0e2352623/sky/backends/cloud_vm_ray_backend.py#L2173-L2199

Note: This PR should be compatible with #1429 as the `validate_launchable` function should only be called by the resources that is meant to be used to create new VMs.

---

This PR also refactors the optimizer logic, so as to make the resource validation clearer.

We previously had two code paths for validating the requested resources causing our implementation for checking needs to be duplicated for requested resources that is launchable and not launchable . https://github.com/skypilot-org/skypilot/blob/15017087357912cfab5a58cdabf9723cfe52a0f3/sky/optimizer.py#L957-L1024

For example, in #2096, although we disable the spot instance in cloud.get_feasible_launchable_resoures and `sky launch --use-spot` will skip the k8s cloud, we can still get around that check by making the requested resources launchable: `sky launch --cloud kubernetes --instance-type 8CPU-16GB --use-spot`.

That said our current code is quite error-prone, making it easy to forget to update the two places for the check.
The refactoring should fix this, by always using the `get_feasible_launchable_resources`

---

Tested:
- [ ] `pytest tests/test_smoke.py`

concretevitamin commented:
Tagging @WoosukKwon @infwinston  - can you take a look as well?

Is there a potential issue where, say K80:4 is the only K80 config available in a region, and users do `sky launch --gpus K80:3 --region ...`. In this case, should we let it succeed (which changes our resource fulfillment semantics to `>=`; currently it is exactly satisfiable)?

Michaelvll commented:
> Is there a potential issue where, say K80:4 is the only K80 config available in a region, and users do `sky launch --gpus K80:3 --region ...`. In this case, should we let it succeed (which changes our resource fulfillment semantics to `>=`; currently it is exactly satisfiable)?

The launch-ability will be checked in the optimizer. The following is our error message for that:
```
> sky launch -c test-gpu --gpus K80:3 --region us-central1 --cloud gcp                    
I 10-16 17:14:56 optimizer.py:880] No resource satisfying {'K80': 3} on [GCP].
I 10-16 17:14:56 optimizer.py:883] Did you mean: ['K80:4', 'K80:8']
sky.exceptions.ResourcesUnavailableError: No launchable resource found for task sky-cmd. To fix: relax its resource requirements.
Hint: 'sky show-gpus --all' to list available accelerators.
      'sky check' to check the enabled clouds.
```

WoosukKwon commented:
Thanks @Michaelvll for quickly fixing the bug! I've checked that this PR resolves the issue.

However, I have a bit fundamental question: do we really need to call `accelerator_in_region_or_zone` (or any other validation check) for `sky exec`? To my understanding, comparing the user's resource requirement with the cluster's resource (by `less_demanding_than`) should be enough for `sky exec`. I believe all the validation checks in `Resources` are only for `sky launch`, and we need to differentiate the two in order to avoid confusion.

Michaelvll commented:
> However, I have a bit fundamental question: do we really need to call `accelerator_in_region_or_zone` (or any other validation check) for `sky exec`? To my understanding, comparing the user's resource requirement with the cluster's resource (by `less_demanding_than`) should be enough for `sky exec`. I believe all the validation checks in `Resources` are only for `sky launch`, and we need to differentiate the two in order to avoid confusion.

Great point @WoosukKwon! The resources should only be checked for availability during launching, instead of creation. I modified the PR thoroughly to move the resource validation to optimization and added several tests to make sure the modification is correct. PTAL.

Tested:
- [ ] `tests/run_smoke_tests.sh`

Michaelvll commented:
This has been fixed in the master branch. Closing this PR for now.

Michaelvll commented:
Sorry, this issue still exists by the following commands:
```
sky gpunode --gpus K80:4 --region us-central1 --cloud gcp -c test-gpu
sky exec --gpus K80:3 --cloud gcp --region us-central1 test-gpu 'echo hi'
```

The error message will be:

```
ValueError: Accelerator "K80" is not available in "us-central1".
```

The error is quite confusing.

Michaelvll commented:
A kind bump for this PR @WoosukKwon

WoosukKwon commented on sky/clouds/service_catalog/__init__.py:
```diff
@@ -1,5 +1,6 @@
 """Service catalog."""
 import collections
+from curses import erasechar
```
It seems this package is not used.

WoosukKwon commented on sky/clouds/service_catalog/gcp_catalog.py:
```diff
@@ -452,9 +393,19 @@ def check_accelerator_attachable_to_host(instance_type: str,
     assert len(acc) == 1, acc
     acc_name, acc_count = acc[0]
 
+    if not instance_type.startswith('n1-'):
+        with ux_utils.print_exception_no_traceback():
+            raise ValueError(
+                f'Accelerator {acc_name} can only be attached to N1 VMs, '
+                f'but {instance_type} is not an N1 VM.')
+
```
We need a special care for A100 GPUs and A2 VMs.
```
$ sky launch --gpus a100 --instance-type a2-highgpu-1g
ValueError: Accelerator A100 can only be attached to N1 VMs, but a2-highgpu-1g is not an N1 VM.
```

WoosukKwon commented on sky/resources.py:
```diff
@@ -709,6 +658,63 @@ def _try_validate_disk_tier(self) -> None:
+            # Validate whether accelerator is available in specified region/zone
+            acc, acc_count = list(acc_requested.items())[0]
+            if self.region is not None or self.zone is not None:
+                if not self._cloud.accelerator_in_region_or_zone(
```
A dumb question: in which case will this branch be taken? I found this error when trying to launch a VM that is not supported in the specified region.

```
$ sky launch --gpus a100:8 --region us-west-1 --cloud aws
I 05-11 13:19:55 optimizer.py:998] No resource satisfying AWS({'A100': 8}) on AWS.
sky.exceptions.ResourcesUnavailableError: No launchable resource found for task Task(run=<empty>)
  resources: AWS({'A100': 8}). Region: us-west-1.
This means the catalog does not contain any instance types that satisfy this request.
To fix: relax or change the resource requirements.
Hint: 'sky show-gpus --all' to list available accelerators.
      'sky check' to check the enabled clouds.
```

WoosukKwon commented on tests/test_cli.py:
```diff
@@ -72,5 +72,5 @@ def _capture_match_gpus_spec(file_path, gpus: str):
         _capture_match_gpus_spec(f.name, 'v100:1')
-        _capture_match_gpus_spec(f.name, 'V100:0.5')
         _capture_match_gpus_spec(f.name, 'V100')
+        _capture_match_gpus_spec(f.name, 'V100:0.5')
```
I feel it's a bit weird that partial GPUs are only supported when the instance type is specified, while they are not supported otherwise. Why don't we disallow partial GPUs for `sky launch`?

```
$ sky launch --gpus v100:0.5
I 05-11 13:23:15 optimizer.py:998] No resource satisfying <Cloud>({'V100': 0.5}) on [AWS, Azure, GCP, Lambda].
I 05-11 13:23:15 optimizer.py:1001] Did you mean: ['V100-32GB:8', 'V100:1', 'V100:2', 'V100:4', 'V100:8']
sky.exceptions.ResourcesUnavailableError: No launchable resource found for task Task(run=<empty>)
  resources: <Cloud>({'V100': 0.5}).
This means the catalog does not contain any instance types that satisfy this request.
To fix: relax or change the resource requirements.
Hint: 'sky show-gpus --all' to list available accelerators.
      'sky check' to check the enabled clouds.

$ sky launch --gpus v100:0.5 --instance-type p3.2xlarge
I 05-11 13:23:22 optimizer.py:636] == Optimizer ==
I 05-11 13:23:22 optimizer.py:647] Target: minimizing cost
I 05-11 13:23:22 optimizer.py:659] Estimated cost: $3.1 / hour
I 05-11 13:23:22 optimizer.py:659] 
I 05-11 13:23:22 optimizer.py:732] Considered resources (1 node):
I 05-11 13:23:22 optimizer.py:781] -----------------------------------------------------------------------------------------
I 05-11 13:23:22 optimizer.py:781]  CLOUD   INSTANCE     vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
I 05-11 13:23:22 optimizer.py:781] -----------------------------------------------------------------------------------------
I 05-11 13:23:22 optimizer.py:781]  AWS     p3.2xlarge   8       61        V100:0.5       us-east-1     3.06          ✔     
I 05-11 13:23:22 optimizer.py:781] -----------------------------------------------------------------------------------------
I 05-11 13:23:22 optimizer.py:781] 
Launching a new cluster 'sky-e1f9-woosuk'. Proceed? [Y/n]:
```

WoosukKwon reviewed (commented):
Thanks @Michaelvll for the effort. Left some comments. Please take a look.

Michaelvll reviewed (commented):
Thank you for the review @WoosukKwon! Based on your comments, I just realized that we have quite many redundancies in the code for checking the instance_type-accelerator combination. I just refactored it. PTAL.

Michaelvll reviewed (commented):
Thank you for the review @WoosukKwon! Based on your comments, I just realized that we have quite many redundancies in the code for checking the instance_type-accelerator combination. I just refactored it. PTAL.

Michaelvll commented:
I further refactored the code in optimizer and clouds to reduce the redundant codes. Please feel free to take another look. @WoosukKwon
Tested:
- [x] `pytest tests/test_optimizer_dryrun.py` with the A100 test included.
- [x] `pytest tests/test_cli.py` with the job execution on existing cluster tests included.
- [ ] `pytest tests/test_smoke.py`

WoosukKwon commented on sky/clouds/cloud.py:
```diff
@@ -470,7 +475,46 @@ def check_disk_tier_enabled(cls, instance_type: str,
+        """
+        assert resources.is_launchable(), resources
+
+        def _equal_accelerators(acc_requested, acc_from_instance_type):
```
Could you annotate the types for the arguments?

WoosukKwon commented on sky/clouds/cloud.py:
```diff
@@ -470,7 +475,46 @@ def check_disk_tier_enabled(cls, instance_type: str,
+            if acc_from_instance_type is None:
+                return False
+
+            for acc in acc_requested:
```
Is this because `acc` is a dictionary?

WoosukKwon commented on sky/clouds/cloud.py:
```diff
@@ -470,7 +475,46 @@ def check_disk_tier_enabled(cls, instance_type: str,
+            for acc in acc_requested:
+                if acc not in acc_from_instance_type:
+                    return False
+                if acc_requested[acc] != acc_from_instance_type[acc]:
+                    return False
```
IIUC, here we are requiring that the requested number of accelerators should match the amount the VM has, right? I think it's worth a comment.

Michaelvll replied:
Good point! Yes, added a docstr for the method.

WoosukKwon commented on sky/clouds/service_catalog/common.py:
```diff
@@ -189,7 +189,7 @@ def _get_all_supported_regions_str() -> str:
     filter_df = df
     if region is not None:
-        filter_df = filter_df[filter_df['Region'].str.lower() == region.lower()]
+        filter_df = _filter_region_zone(filter_df, region, None)
```
Style nit: IIRC, the convention is to use kwarg when the provided argument is literal, isn't it?
```suggestion
        filter_df = _filter_region_zone(filter_df, region, zone=None)
```
Same for L579 and L594.

Michaelvll replied:
Good point! Fixed. Thanks!

WoosukKwon commented on tests/test_cli.py:
```diff
@@ -4,6 +4,8 @@
 
 import sky
+from sky import exceptions
+from sky import clouds
```
```suggestion
```

Michaelvll replied:
Oops, good catch! Fixed. Thanks!

WoosukKwon commented on tests/test_jobs.py:
```diff
@@ -0,0 +1,123 @@
+            db_utils.SQLiteConn(str(db_path), global_user_state.create_table))
+
+    @pytest.fixture
+    def _mock_cluster_state(self, _mock_db_conn, enable_all_clouds):
+        assert 'state.db' not in global_user_state._DB.db_path
+
```
Can you write a doctoring or comment that briefly describes the resources that the three clusters have? It'll be very helpful for quickly understanding the test code below.

Michaelvll replied:
Good point! Added. Thanks!

WoosukKwon reviewed (approved):
LGTM. Thanks for the PR.

Michaelvll commented:
Thanks a lot for the review @WoosukKwon!! I am testing the smoke tests before merging:
- [ ] `pytest tests/test_smoke.py`
